internal/app/machined/pkg/controllers/kubeaccess/talossa/crd_controller.go

```diff
@@ -0,0 +1,633 @@
+	resourceInformer := dynamicInformerFactory.ForResource(talosSAGVR)
+	informer := resourceInformer.Informer()
+
+	indexer := informer.GetIndexer()
+	lister := dynamiclister.New(indexer, talosSAGVR)
+
+	kubeInformerFactory := kubeinformers.NewSharedInformerFactory(kubeCli, informerResyncPeriod)
+	secrets := kubeInformerFactory.Core().V1().Secrets()
+
+	logger.Debug("creating event broadcaster")
+
+	eventBroadcaster := record.NewBroadcaster()
+	eventBroadcaster.StartStructuredLogging(0)
+	eventBroadcaster.StartRecordingToSink(&typedcorev1.EventSinkImpl{Interface: kubeCli.CoreV1().Events("")})
+
+	recorder := eventBroadcaster.NewRecorder(scheme.Scheme, corev1.EventSource{Component: controllerAgentName})
+
+	controller := CRDController{
+		talosCA:                talosCA,
+		allowedNamespaces:      allowedNamespaces,
+		allowedRoles:           slices.ToSet(allowedRoles),
+		dynamicInformerFactory: dynamicInformerFactory,
+		kubeInformerFactory:    kubeInformerFactory,
+		kubeClient:             kubeCli,
+		dynamicClient:          dynCli,
+		dialer:                 dialer,
+		dynamicLister:          lister,
+		queue: workqueue.NewNamedRateLimitingQueue(
+			workqueue.DefaultControllerRateLimiter(),
+			constants.ServiceAccountResourceKind,
+		),
+		logger:         logger,
+		secretsSynced:  secrets.Informer().HasSynced,
+		talosSAsSynced: informer.HasSynced,
+		eventRecorder:  recorder,
+		secretsLister:  secrets.Lister(),
+	}
+
+	informer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: controller.enqueueTalosSA,
+		UpdateFunc: func(oldTalosSA, newTalosSA interface{}) {
+			controller.enqueueTalosSA(newTalosSA)
+		},
+	})
+
+	secrets.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: controller.handleSecret,
+		UpdateFunc: func(oldSec, newSec interface{}) {
+			newSecret := newSec.(*corev1.Secret) //nolint:errcheck
+			oldSecret := oldSec.(*corev1.Secret) //nolint:errcheck
+
+			if newSecret.ResourceVersion == oldSecret.ResourceVersion {
+				return
+			}
+
+			controller.handleSecret(newSec)
+		},
+		DeleteFunc: controller.handleSecret,
+	})
+
+	return &controller, nil
+}
+
+// Run starts the CRD controller.
+func (t *CRDController) Run(ctx context.Context, workers int) error {
+	defer func() {
+		t.queue.ShutDown()
+		t.dialer.CloseAll()
+	}()
+
+	t.kubeInformerFactory.Start(ctx.Done())
+	t.dynamicInformerFactory.Start(ctx.Done())
+
+	t.logger.Sugar().Debugf("starting %s controller", constants.ServiceAccountResourceKind)
+
+	t.logger.Debug("waiting for informer caches to sync")
+
+	if ok := cache.WaitForCacheSync(ctx.Done(), t.secretsSynced, t.talosSAsSynced); !ok {
+		return fmt.Errorf("failed to wait for caches to sync")
+	}
+
+	t.logger.Debug("starting workers")
+
+	for i := 0; i < workers; i++ {
+		go wait.Until(func() {
+			t.runWorker(ctx)
+		}, time.Second, ctx.Done())
+	}
+
+	t.logger.Debug("started workers")
+	<-ctx.Done()
+	t.logger.Debug("shutting down workers")
```

Comment:
it doesn't look like we wait here for the workers to stop? (or does it happen implicitly via something else?)

Reply:
Yep, controller is stopped when the context passed to `.Run` is cancelled. Context cancellation causes deferred `t.queue.ShutDown()` on line 203 to run, which causes the loop `for t.processNextWorkItem(ctx)` on line 234 to terminate, which terminates the worker. I also did test locally to make sure that the workers actually are stopped.

Reply:
so my concern would be that we don't wait for the workers to  stop... so if they don't actually stop, we would never notice that and it might leave some extra/stale workers running.

Reply:
I added a wait using a waitgroup and locally verified that workers actually terminate and we wait for them. Feel free to have another look.